Add any_object type-erasing wrapper

lewissbaker opened the pull request:
Adds a move-only `unifex::any_object` type-erasing wrapper that supports the small-object optimisation.

Also refactors some of the type-erasing helper classes from `any_unique` to make them more reusable.

Some questions on semantics:
* If I pass an allocator with `std::allocator_arg` to the constructor but the type to be constructed fits in the small-object buffer then should it use the allocator or should it still use the small-object buffer, ignoring the allocator (currently does the latter)
* Should `any_object` allow non-movable objects that would be heap-allocated because it's a large object (move ctor is not called on concrete type in this case)?
* If so, should `any_object` also store small objects that are not movable in a heap-allocation?
* Or should we just always require that wrapped types are always move-constructible? (this is the current behaviour)
* Are we ok with requiring that destructors are noexcept?
* `any_object` currently utilises ADL-isolation techniques but this means we can't deduce template args for this type. Is this going to break any use-cases for writing customisations of algorithms for `any_object`?
* Are there better ways of structuring the template args to make them more ergonomic?
  Currently we have `any_object<InlineSize, InlineAlignment, RequireNoExceptMove, DefaultAllocator, CPOs...>`
* Should the move constructor leave the source object in a valid but moved from state when moving a small object? Or should it destroy the source object after a successful move?

TODO
- [ ] Add more tests for any untested use-cases
- [ ] Write documentation for type-erasing wrappers (`any_unique` as well as `any_object`)

ispeters commented on include/unifex/detail/with_type_erased_tag_invoke.hpp:
```diff
@@ -0,0 +1,123 @@
+      void* objPtr = get_object_address(t);
+      auto* fnPtr = get_vtable(t)->template get<CPO>();
+
+      // Sanity-check noexceptness of implementation.
```
I like that the comment on the sanity check in the forwarding implementation includes the justification for the sanity check; consider copying it here.

ispeters commented on include/unifex/overload.hpp:
```diff
@@ -64,9 +61,14 @@ struct _sig {};
 template <auto& CPO, typename Sig>
 using overload_t = typename _overload::_cpo_t<tag_t<CPO>, Sig>::type;
 
+
+
```
Delete?
```suggestion
```

ispeters commented on test/any_object_test.cpp:
```diff
@@ -0,0 +1,295 @@
+
+namespace
+{
+  inline constexpr struct get_typeid_cpo {
+    using type_erased_signature_t =
+        unifex::type_index(const unifex::this_&) noexcept;
+
+    template <typename T>
+    friend unifex::type_index tag_invoke(get_typeid_cpo, const T& x) {
+      return unifex::type_id<T>();
+    }
+
+    template(typename T)                                     //
+        (requires unifex::tag_invocable<get_typeid_cpo, T>)  //
+        auto
+        operator()(const T& x) const noexcept
+        -> unifex::tag_invoke_result_t<get_typeid_cpo, const T&> {
+      static_assert(std::is_same_v<
+                    unifex::type_index,
+                    unifex::tag_invoke_result_t<get_typeid_cpo, const T&>>);
+      return unifex::tag_invoke(get_typeid_cpo{}, x);
+    }
+  } get_typeid{};
```
This looks familiar; is there already a similar thing in another test?  Might be worth having a "test utils" directory for stuff like this.

lewissbaker replied:
I was wondering if we should just add this as a first-class CPO to the library itself.
I can see potential utility for it - e.g. to provide something similar to the `std::any::type()` method to get the `std::type_info` of the current object.

ispeters reviewed (approved):
Looks good.  Regarding your questions, here are my answers:
> If I pass an allocator with `std::allocator_arg` to the constructor but the type to be constructed fits in the small-object buffer then should it use the allocator or should it still use the small-object buffer, ignoring the allocator (currently does the latter)

I like the current choice.

> Should `any_object` allow non-movable objects that would be heap-allocated because it's a large object (move ctor is not called on concrete type in this case)?  If so, should any_object also store small objects that are not movable in a heap-allocation?  Or should we just always require that wrapped types are always move-constructible? (this is the current behaviour)

Do you have some scenarios in mind to help think this one through?  I think my knee-jerk response is:
 - yes, we should support immovable types, and
 - yes, we should heap-allocate small immovable objects.

My usual work-around to passing immovable things around is to pass `unique_ptr<T>` instead of `T` so if there's a real scenario where I want to pass around an `any_object` that wraps an immovable `T`, I think it's unsurprising that the result is a heap allocation.

> Are we ok with requiring that destructors are noexcept?

I am, but perhaps noisily (i.e. a `static_assert` that the destructor is `noexcept` so that we don't silently introduce an unexpected termination path).

> `any_object` currently utilises ADL-isolation techniques but this means we can't deduce template args for this type. Is this going to break any use-cases for writing customisations of algorithms for `any_object`?

I don't know.

> Are there better ways of structuring the template args to make them more ergonomic?  Currently we have any_object<InlineSize, InlineAlignment, RequireNoExceptMove, DefaultAllocator, CPOs...>

I expect it'll be more common to have a reason to customize the CPOs than the other parameters; perhaps there should be a type `cpos_t` (analogous to `tag_t`) that is like a type list of CPOs, so that you can put the CPO pack first, and then provide some sensible defaults for the other parameters:
```
template <
    template <typename... CPOs> typename CPOList,
    std::size_t InlineSize = sizeof(void*),
    std::size_t InlineAlignment = alignof(void*),
    typename DefaultAllocator = std::allocator<char>,
    bool RequireNoExceptMove = true>
```

> Should the move constructor leave the source object in a valid but moved from state when moving a small object? Or should it destroy the source object after a successful move?

Perhaps this should be like `std::string`: moved-from large objects are `nullptr` and moved-from small objects contain a moved-from `T`.

lewissbaker commented on include/unifex/any_object.hpp:
```diff
@@ -0,0 +1,200 @@
+      DefaultAllocator,
+      CPOs...>::type : private with_type_erased_tag_invoke<type, CPOs>... {
+    // Pad size/alignment out to allow storage of at least a pointer.
+    static constexpr std::size_t padded_alignment =
+        InlineAlignment < alignof(void*) ? alignof(void*) : InlineAlignment;
+    static constexpr std::size_t padded_size =
+        InlineSize < sizeof(void*) ? sizeof(void*) : InlineSize;
```
I originally did use `std::max` here which is constexpr since C++14 but was concerned about the overhead of pulling in the entire `<algorithm>` header just to be able to use it. Maybe we can put our own constexpr `max` in the `unifex` namespace somewhere?

I also went through various iterations of trying to pad the padded_size up to the next multiple of `padded_alignment` but wasn't sure whether or not that was worthwhile or not. It might take up extra padding space that would otherwise be used by classes that inherit from this one.

ispeters replied:
A `unifex::max` seems like a good balance between expressing intent and minimizing include overhead.

lewissbaker replied:
Any thoughts on where such a utility should live in libunifex?

ispeters replied:
We have `<unifex/std_concepts.hpp>`; perhaps there should be a `<unifex/std_algorithms.hpp>`?

lewissbaker commented on include/unifex/any_object.hpp:
```diff
@@ -0,0 +1,200 @@
+  };
+
+  template <
+      std::size_t InlineSize,
+      std::size_t InlineAlignment,
+      bool RequireNoexceptMove,
+      typename DefaultAllocator,
```
One thing I was considering is having `basic_any_object` with all these parameters and then having `any_object` be an alias that applies those defaults.

Although I'm less certain on what the default small-object buffer size should be. A pointer storage won't waste any space but probably won't apply the SOO to many types.

ispeters replied:
Perhaps the default should be `sizeof(std::string)` (either literally, on the assumption that platform-specific sizes for strings might be chosen for relevant reasons, or figuratively, by which I mean "the size of a string you have handy for measuring").

lewissbaker replied:
I've just put the default as `3 * sizeof(void*)` at this stage so that the overall object size is 4 x pointers (16 or 32 bytes depending on arch).

lewissbaker commented on include/unifex/any_ref.hpp:
```diff
@@ -0,0 +1,104 @@
+
+  // Two any_ref's compare equal IFF they refer to the same object (shallow).
+  //
+  // TODO: This is not entirely accurate as it might be possible to have two
+  // any_ref objects constructed from the same object at the same address but
+  // with different static-types (e.g. one any_ref is constructed from a base-class)
+  // and thus they would have different vtables. However, we can't just compare
+  // vtables since it's not guaranteed that function pointers will point to the
+  // same address.
```
I guess it depends what we want "equality" to mean for type-erased types.
For other type-erasing wrappers it will generally check that the typeid is the same and that `operator==()` returns `true`.

With `any_ref` it's possible to construct two different views onto the same object that have vtables generated for different types and this might cause the `any_ref` objects to behave differently - they might dispatch to different static implementations.

Also, defining `any_ref` purely in terms of object identity means that it won't return `true` for cases where two different concrete objects of the same type that compare equal are wrapped in `any_ref` objects and you compare those `any_ref` objects.

e.g. do we want this to pass?
```
concrete a = ...;
concrete b = ...;
assert(a == b);
any_ref<foo_cpo> aRef = a;
any_ref<foo_cpo> bRef = b;
assert(aRef == bRef);
```

ispeters replied:
Thanks for that elaboration; it makes the question clearer, and spurred a thought: perhaps there should be a well-defined, low-friction way to opt into a virtual implementation of equality.

If you provide it, you get deep equality, and if you don't, you get pointer equality.

I like the pay-for-what-you-use nature of this proposal, but I wonder what it costs in terms of complexity, or the evils of making everything configurable.

lewissbaker commented on include/unifex/detail/any_heap_allocated_storage.hpp:
```diff
@@ -0,0 +1,161 @@
+      struct state {
+        template(typename... Args)                     //
+            (requires constructible_from<T, Args...>)  //
+            explicit state(
+                std::allocator_arg_t,
+                allocator_type allocator,
+                std::in_place_type_t<T>,
+                Args&&... args)
```
That's what clang-format seems to do to code after the `template` and `requires` macros - seems to add an extra level of indentation.

lewissbaker commented on include/unifex/detail/any_heap_allocated_storage.hpp:
```diff
@@ -0,0 +1,161 @@
+                Args&&... args) {
+          state_ = allocator_traits::allocate(allocator, 1);
+
+          auto deallocateOnExit = scope_guard{[&]() noexcept {
+            allocator_traits::deallocate(allocator, state_, 1);
+          }};
+
+          allocator_traits::construct(
+              allocator,
+              state_,
+              std::allocator_arg,
+              allocator,
+              std::in_place_type<T>,
+              static_cast<Args&&>(args)...);
+
+          deallocateOnExit.release();
```
As far as I can see under -fno-exceptions or if the constructor is declared noexcept all of this just optimises out. Even when exceptions are possible,  there isn't any storage of the flag or branching on it.

lewissbaker commented on include/unifex/detail/any_heap_allocated_storage.hpp:
```diff
@@ -0,0 +1,161 @@
+                std::in_place_type<T>,
+                static_cast<Args&&>(args)...) {}
+
+        // Ideally this would only be instantiated if it's actually called.
+        //
+        // This should have a 'requires std::copy_constructible<T>' on the
+        // definition but we can't currently do that with the concepts emulation
+        // layer so we'll just enable the check when we actually have concepts.
+        base(const base& other)
+#if UNIFEX_CXX_CONCEPTS
+            requires copy_constructible<T>
+#endif
+          : base(
```
Yes, even if you have a template function that would match the signature of a copy-constructor, this doesn't prevent the compiler for generating a default copy-constructor for you and calling that instead.

So we need to explicitly declare a copy-constructor here to prevent the compiler for generating a default one.

lewissbaker commented on include/unifex/detail/type_erasure_builtins.hpp:
```diff
@@ -0,0 +1,83 @@
+// copy/move constructors.
+//
+// These CPOs can be passed to the list of CPOs for a vtable<> to create vtable
+// entries for these operations. These CPOs are only customisable through the
+// builtin operations and not via tag_invoke().
```
You customise the `destroy_cpo` by defining a destructor.
You customise the `move_construct_cpo` by defining a move constructor.
You customise the `copy_construct_cpo` by defining a copy-constructor.

ispeters replied:
Ok, then I suggest rewording this part
> These CPOs are only customisable through the builtin operations and not via tag_invoke().

to say "special member functions" instead of "builtin operations", and/or just copy your comment above into the source.

lewissbaker commented on include/unifex/detail/type_erasure_builtins.hpp:
```diff
@@ -0,0 +1,83 @@
+      }
+    };
+
+    template <bool RequireNoexceptMove>
+    struct _move_construct_cpo {
+      using type_erased_signature_t =
+          void(void* p, this_&& src) noexcept(RequireNoexceptMove);
+
+      template(typename T)  //
+          (requires(
+              !RequireNoexceptMove ||
+              std::is_nothrow_move_constructible_v<T>))  //
+          void
+          operator()(void* p, T&& src) const noexcept(RequireNoexceptMove) {
+        ::new (p) T(static_cast<T&&>(src));
+      }
+    };
+
+    template <bool RequireNoexceptCopy>
+    struct _copy_construct_cpo {
+      using type_erased_signature_t =
+          void(void* p, const this_& src) noexcept(RequireNoexceptCopy);
+
+      template(typename T)  //
+          (requires(
+              !RequireNoexceptCopy ||
+              std::is_nothrow_copy_constructible_v<T>))  //
+          void
+          operator()(void* p, const T& src) const
+          noexcept(RequireNoexceptCopy) {
+        ::new (p) T(src);
+      }
+    };
```
Yes, I think you're right in that we should be able to declare the `operator()` function as `noexcept(std::is_nothrow_copy_constructible_v<T>)` instead of `noexcept(RequiresNoexceptCopy)`.

It shouldn't hurt for the type-erased thunk (defined in vtable.hpp) that dispatches to this CPO may be declared `noexcept(false)` but it should still be fine to call a noexcept function from this.

lewissbaker commented:
> > Should `any_object` allow non-movable objects that would be heap-allocated because it's a large object (move ctor is not called on concrete type in this case)?  If so, should any_object also store small objects that are not movable in a heap-allocation?  Or should we just always require that wrapped types are always move-constructible? (this is the current behaviour)
> 
> Do you have some scenarios in mind to help think this one through?

One concern might be that if I have defined my `any_object` type with inline size/alignment using `sizeof(T)` and `alignof(T)` to ensure that it is able to store type `T` in-place in the object but then `T` happens to not be move-constructible and instead of getting a compile-error it will silently still accept `T` but heap-allocate it instead.

The other more subtle variant of this may occur in the case where we have specified `RequireNoexceptMove = true` and we try to wrap a type, `T` that is move-constructible but is not nothrow-move-constructible. Should this case silently perform a heap-allocation instead of raising a compile error? I'm not sure whether we should be targeting "make it work in as many places as possible" or "don't silently give the user performance footguns". I guess  you're leaning more towards the former.

We could also add an `any_immovable` that is like `any_object` but without the move-constructor support for non-movable types like operation-states.

> I think my knee-jerk response is:
> 
>     * yes, we should support immovable types, and
> 
>     * yes, we should heap-allocate small immovable objects.
> 
> My usual work-around to passing immovable things around is to pass `unique_ptr<T>` instead of `T` so if there's a real scenario where I want to pass around an `any_object` that wraps an immovable `T`, I think it's unsurprising that the result is a heap allocation.

> > Are we ok with requiring that destructors are noexcept?
> 
> I am, but perhaps noisily (i.e. a `static_assert` that the destructor is `noexcept` so that we don't silently introduce an unexpected termination path).

The current behaviour will be that you'll get an error saying that `any_object` is not constructible from `foo` if `foo` as a potentially-throwing destructor.

> > Are there better ways of structuring the template args to make them more ergonomic?  Currently we have any_object<InlineSize, InlineAlignment, RequireNoExceptMove, DefaultAllocator, CPOs...>
> 
> I expect it'll be more common to have a reason to customize the CPOs than the other parameters; perhaps there should be a type `cpos_t` (analogous to `tag_t`) that is like a type list of CPOs, so that you can put the CPO pack first, and then provide some sensible defaults for the other parameters:
> 
> ```
> template <
>     template <typename... CPOs> typename CPOList,
>     std::size_t InlineSize = sizeof(void*),
>     std::size_t InlineAlignment = alignof(void*),
>     typename DefaultAllocator = std::allocator<char>,
>     bool RequireNoExceptMove = true>
> ```

As mentioned elsewhere, we could rename the current alias to `basic_any_object` and then define `any_object` as an alias with the defaults you suggest (although possibly with a slightly larger small-object buffer - maybe so the total size of `any_object` is 4 pointers?)

> > Should the move constructor leave the source object in a valid but moved from state when moving a small object? Or should it destroy the source object after a successful move?
> 
> Perhaps this should be like `std::string`: moved-from large objects are `nullptr` and moved-from small objects contain a moved-from `T`.

This is the current behaviour.

If we were to actually eagerly destroying the source object would require replacing the vtable with a dummy vtable whose destructor entry (and maybe move-construct entry) was a no-op. Other entries would probably need to just call std::terminate.

Another question: Should `any_object` also support move-assignment?
If so this potentially also brings up the potential for leaving the object in an invalid state if we fail to construct the new object after having destroyed the previous object.

lewissbaker commented on include/unifex/any_object.hpp:
```diff
@@ -0,0 +1,200 @@
+    }
+
+    vtable_holder_t vtable_;
+    alignas(padded_alignment) std::byte storage_[padded_size];
```
This could just as easily be `unsigned char` if `std:byte` is going to cause issues.

ispeters replied:
Yeah, we're stuck with a C++17 compiler and a not-quite-C++14 stdlib; it'd be easier to integrate this if you avoided `std::byte`.

ispeters commented:
> One concern might be that if I have defined my `any_object` type with inline size/alignment using `sizeof(T)` and `alignof(T)` to ensure that it is able to store type `T` in-place in the object but then `T` happens to not be move-constructible and instead of getting a compile-error it will silently still accept `T` but heap-allocate it instead.
>
> The other more subtle variant of this may occur in the case where we have specified `RequireNoexceptMove = true` and we try to wrap a type, `T` that is move-constructible but is not nothrow-move-constructible. Should this case silently perform a heap-allocation instead of raising a compile error? I'm not sure whether we should be targeting "make it work in as many places as possible" or "don't silently give the user performance footguns". I guess you're leaning more towards the former.
>
> We could also add an `any_immovable` that is like `any_object` but without the move-constructor support for non-movable types like operation-states.

Maybe I'm misjudging here, but it seems like the concerns you're raising stem from the perspective of someone trying to type-erase a thing while still knowing exactly which type it is, and I'm not sure that's a common case.  Do you think it is?

For example, the `any_scheduler` that @ericniebler added to enable scheduler affinity in `task<>` can't be optimized for any particular scheduler.  A function that receives an `any_*` is, presumably, doing so to create an interface boundary where the caller doesn't have to expand library templates.  In cases like these, where a library wants to handle user-provided objects with type erasure, "make it work in as many places as possible" seems valid; naively, the alternative seems like "throw up what will feel like arbitrary barriers to integration".

Perhaps a place where you'd like to ensure that your `any_*` is sized _exactly_ right is when you want to _return_ the `any_*` from a function at an interface boundary.  Could we serve that use case with a `constexpr` predicate that you could use to `static_assert` that `any_*<...>` is "ideal" for `T`?  Then you could declare your `any_*` in the header, but `static_assert(unifex::is_ideally_sized_any<T, my_any>)` in the implementation, where the details are known.  We could then write `is_ideally_sized_any` so that the ideal size for immovable objects is "just enough for a pointer".

> Another question: Should `any_object` also support move-assignment?
> If so this potentially also brings up the potential for leaving the object in an invalid state if we fail to construct the new object after having destroyed the previous object.

Can't you support move-assignment without that risk like so:
```
any_object& operator=(any_object&& rhs) noexcept(is_nothrow_move_constructible<any_object>::value) {
  any_object tmp(std::move(rhs));
  swap(tmp);
  return *this;
}
```

lewissbaker commented:
> Maybe I'm misjudging here, but it seems like the concerns you're raising stem from the perspective of someone trying to type-erase a thing while still knowing exactly which type it is, and I'm not sure that's a common case. Do you think it is?

I think there are definitely cases where we want to make sure that certain concrete types do not allocate when type-erased.
e.g. where we have a `virtual` method returning an `any_sender` and we want to make sure that the known (or possibly all) overrides of this method return types that are stored inline.

> For example, the `any_scheduler` that @ericniebler added to enable scheduler affinity in `task<>` can't be optimized for any particular scheduler. A function that receives an `any_*` is, presumably, doing so to create an interface boundary where the caller doesn't have to expand library templates. In cases like these, where a library wants to handle user-provided objects with type erasure, "make it work in as many places as possible" seems valid; naively, the alternative seems like "throw up what will feel like arbitrary barriers to integration".

The `any_scheduler` example is perhaps not the ideal example here as the scheduler is required to be move-constructible and so would not run into this issue, but I understand the point you're making.

I guess this is just a question of high-level priorities: do we want to prefer making things work if we can but with the possibility of silently introducing some inefficiencies? Or do we want to try to prevent the user from running into unexpected perf problems at the expense of putting up some 

> Perhaps a place where you'd like to ensure that your `any_*` is sized _exactly_ right is when you want to _return_ the `any_*` from a function at an interface boundary. Could we serve that use case with a `constexpr` predicate that you could use to `static_assert` that `any_*<...>` is "ideal" for `T`? Then you could declare your `any_*` in the header, but `static_assert(unifex::is_ideally_sized_any<T, my_any>)` in the implementation, where the details are known. We could then write `is_ideally_sized_any` so that the ideal size for immovable objects is "just enough for a pointer".

Yes, this is one example of where I'd like to be able to provide guarantees about type-erasure not heap-allocating.
Perhaps `any_*` types that have hybrid storage strategies could provide an `any_*::is_stored_inline_v<T>` query.
Or maybe we could add another constructor that you pass an extra tag type that forces this check for you at the call-site.

e.g. `my_any{std::in_place, std::in_place_type<T>, args...)`

Another strategy might be to pass a dummy allocator type that raises a compile error if you ever try to call `allocate()`.
`my_any(std::allocator_arg, disallow_allocation, std::in_place_type_t<T>, args...)`?

> > Another question: Should `any_object` also support move-assignment?
> > If so this potentially also brings up the potential for leaving the object in an invalid state if we fail to construct the new object after having destroyed the previous object.
> 
> Can't you support move-assignment without that risk like so:
> 
> ```
> any_object& operator=(any_object&& rhs) noexcept(is_nothrow_move_constructible<any_object>::value) {
>   any_object tmp(std::move(rhs));
>   swap(tmp);
>   return *this;
> }
> ```

The problem here is that the default implementation of `swap()` is defined in terms of move-assignment, so this implementation would just infinitely recurse. Even if we did have `any_object` customise `swap()` it would still need to deal with the case where both objects are different types and stored in-place.

The target of the assignment needs to destroy the existing object first and then try to move-construct the source object into the target storage. If that throws then the target object is left without a valid object and so would need to populate it with a vtable that has a no-op destroy_cpo entry, and maybe entries for all other CPOs that just throws an exception (like `std::visit` throws `std::bad_variant_access`).

I think you'd need to do something like:
```c++
any_object& operator=(any_object&& rhs) noexcept(RequireNoexceptMove) {
  auto* destroyFn = vtable_.template get<destroy_cpo>();
  destroyFn(&storage_);
  if constexpr (!RequreNoexceptMove) {
    // If move ctor throws then leave in a state so that destroy() is a no-op.
    vtable_ = vtable_holder_t::create_empty();
  }
  auto* moveFn = rhs.vtable_.template get<move_construct_cpo<RequireNoexceptMove>>();
  moveFn(&storage_, &rhs.storage_); 
  vtable_ = rhs.vtable_;
  return *this;
}
```

ispeters commented:
> I guess this is just a question of high-level priorities: do we want to prefer making things work if we can but with the possibility of silently introducing some inefficiencies? Or do we want to try to prevent the user from running into unexpected perf problems at the expense of putting up some

(I wonder what you intended to finish that sentence with.)

> Yes, this is one example of where I'd like to be able to provide guarantees about type-erasure not heap-allocating.
Perhaps `any_*` types that have hybrid storage strategies could provide an `any_*::is_stored_inline_v<T>` query.
Or maybe we could add another constructor that you pass an extra tag type that forces this check for you at the call-site.
>
> e.g. `my_any{std::in_place, std::in_place_type<T>, args...)`
>
> Another strategy might be to pass a dummy allocator type that raises a compile error if you ever try to call allocate().
`my_any(std::allocator_arg, disallow_allocation, std::in_place_type_t<T>, args...)`?

We're imagining a case where a library user wants to pass around some object in a type-erased way, but the concrete type is immovable.  Suppose we throw up barriers to this in an effort to prevent accidental inefficient memory usage—how does the user achieve their goal?  They still want to pass around a type-erased immovable object.  Do we make them write a handle type (a struct around a `unique_ptr<T>`, say, with all the relevant CPOs forwarded) and type-erase the handle?  That seems like a lot of ceremony for what _could_ be solved in the `any_*` implementation by just choosing the heap-allocated storage strategy.

Perhaps, instead of "an extra tag type that forces this check for you at the call-site", the "default to performant code" strategy would be a tag that does the inverse: a tag that permits type erasing an immovable type:

`my_any{unifex::in_place_immovable_type<T>, args...}`

I could see the constructor overload that takes a `std::in_place_type<T>` containing this:
```
static_assert(
    std::is_move_constructible_v<T>,
    "Consider the overload with unifex::in_place_immovable_type<T>; "
    "note that immovable objects will be heap-allocated and therefore don't need much inline storage space.");
```

> The problem here is that the default implementation of `swap()` is defined in terms of move-assignment…

🤦🏼‍♂️ Right.

> The target of the assignment needs to destroy the existing object first and then try to move-construct the source object into the target storage. If that throws then the target object is left without a valid object and so would need to populate it with a vtable that has a no-op destroy_cpo entry, and maybe entries for all other CPOs that just throws an exception (like `std::visit` throws `std::bad_variant_access`).

I'm not sure you can substitute "just throw an exception" for the other CPOs since they could easily be `noexcept`.  I think the "I'm broken" vtable would have to terminate in the unimplemented CPOs.

What if `any_object` only supported move-assignment when `RequireNoexceptMove` is true?

ericniebler commented on include/unifex/type_traits.hpp:
```diff
@@ -150,6 +150,33 @@ using decay_rvalue_t =
+inline constexpr bool is_nothrow_convertible_v = std::is_nothrow_convertible_v<From, To>;
+
+template<typename From, typename To>
+using is_nothrow_convertible = std::is_nothrow_convertible<From, To>;
```
Personally, I would make these using declarations.

ericniebler commented on include/unifex/any_unique.hpp:
```diff
@@ -492,12 +210,6 @@ using any_unique = typename _any_unique::_byval<CPOs...>::type;
-
 } // namespace unifex
 
 #include <unifex/detail/epilogue.hpp>
```
Is it not possible to implement `any_unique` in terms of `basic_any_object`?

lewissbaker replied:
I guess in theory, `any_unique` could be implemented in terms of `basic_any_object` with an inline-size of 0.

There might be some small perf difference, though, in that with any_unique we know that the data pointer is always a pointer and so the caller of a CPO can load the data pointer in a register and pass that directly (possibly avoiding needing to spill the any_unique object to the stack). Whereas with any_object, since the object might be stored inline we have to pass the address of the data pointer into the CPO function which then interprets that memory as a pointer and loads the address of the object. This seems more likely to require the compiler to write the type-erasing wrapper to memory.

This is all just a hunch, though. I've not benchmarked or measured the runtime performance difference here....

ericniebler reviewed (commented):
Mostly minor tweaks here and there. Overall, looks great.

ericniebler commented:
> * If I pass an allocator with std::allocator_arg to the constructor but the type to be constructed fits in the small-object buffer then should it use the allocator or should it still use the small-object buffer, ignoring the allocator (currently does the latter)
Should any_object allow non-movable objects that would be heap-allocated because it's a large object (move ctor is not called on concrete type in this case)?

I agree it should ignore the allocator, except to use the allocator to do the inplace construction and destruction. (I don't know why that's a thing but it is.)

> * If so, should any_object also store small objects that are not movable in a heap-allocation?
Or should we just always require that wrapped types are always move-constructible? (this is the current behaviour)

I agree with requiring move construction.

> * Are we ok with requiring that destructors are noexcept?

Absolutely.

> * any_object currently utilises ADL-isolation techniques but this means we can't deduce template args for this type. Is this going to break any use-cases for writing customisations of algorithms for any_object?

In my prototype for P2300, I used a different expression of the ADL isolation technique. Rather than `foo<Args...>::type`, I used `foo<id<Args>...>`. Seems to work well. Maybe libunifex should switch. That would be a large undertaking, though.


> * Are there better ways of structuring the template args to make them more ergonomic?

> * Currently we have any_object<InlineSize, InlineAlignment, RequireNoExceptMove, DefaultAllocator, CPOs...>
Should the move constructor leave the source object in a valid but moved from state when moving a small object? Or should it destroy the source object after a successful move?

lewissbaker commented on include/unifex/swap.hpp:
```diff
@@ -92,21 +82,21 @@ namespace _swap {
 inline constexpr _swap::_fn swap {};
 } // namespace _swap_cpo
-using _swap_cpo::swap;
+using namespace _swap_cpo;
```
I'm pretty sure that the using _swap_cpo::swap; introduces swap explicitly as an object which can then conflict with friend-functions declared on types in the same namespace.
However, with the 'using namespace' approach it doesn't introduce a conflict.
I think all CPOs that are implemented in terms of an ADL-name that is the same as the CPO name need to be imported either as inline namespaces (with implicit 'using namespace') or explicit 'using namespace'.

See https://github.com/facebookexperimental/libunifex/pull/304/commits/1e067b1a4f7bfd5c1062dccd05a8a265662c5909

lewissbaker commented on doc/type_erasure.md:
```diff
@@ -0,0 +1,590 @@
+* **VTable Storage** - How and where is the VTable for the type-erased object stored?
+  * **Indirect VTable** - Wrapper stores a pointer to the statically allocated vtable. Pointer indirection to call the method.
+  * **In-Place VTable** - Wrapper stores vtable entries in-line inside the wrapper.
+    Avoids pointer-indirection at cost of larger wrapper types. Also allows casting other type-erased types to wrapper types with narrower types.
```
The mention of casting here was intended to refer to things like casting `any_object<cpo1, cpo2, cpo3>`  to `any_object<cpo1, cpo3>`
This cannot in general be supported using indirect vtables because the layout/ordering of the vtable entries is not necessarily compatible.

ericniebler replied:
Right. That's true. I was talking about whether we want to support "base" and "derived" type-erasing wrappers, where the derived type adds its entries to the vtable after the base entries. `folly::Poly` supports that, and automatically allows derived-to-base conversions.

lewissbaker replied:
This is not a use-case I've considered. I'll think about how to address this but will leave it for another PR.

lewissbaker commented on doc/type_erasure.md:
```diff
@@ -0,0 +1,590 @@
+  * **Hybrid VTable** - Wrapper stores vtable inline for small vtables and stores it out-of-line for large vtables.
+* **Copyability** - Whether the wrapper is copyable, move-only or non-movable.
+  * **Copyable** - Wrapper has a copy-constructor and move-constructor (may be either `noexcept(true)` or `noexcept(false)`)
+  * **Move-Only** - Wrapper has move-constructor but no copy-constructor (my be either `noexcept(true)` or `noexcept(false)`)
```
Where does this leave types that implement a throwing copy constructor (e.g. because it allocates some resource) but that do not have a move-constructor implemented and so the move-constructor falls back to the throwing copy-constructor?

> When dynamic allocation is allowed, types with throwing move constructors should live on the heap.

This is a discussion that I was having with @ispeters before.
When move-constructors are required to be nothrow is it preferable to not support types that don't have noexcept move constructors or to paper over their lack of nothrow move ctor by heap-allocating them?
I guess it seems like you are on the side of "make it work even if it's at the cost of a heap allocation".

ericniebler replied:
> Where does this leave types that implement a throwing copy constructor (e.g. because it allocates some resource) but that do not have a move-constructor implemented and so the move-constructor falls back to the throwing copy-constructor?

You mean buggy types. I don't care to support buggy types. Fix your types.

lewissbaker replied:
Ok, I've updated it so that it heap allocates the value if either the concrete type is not nothrow-move-constructible or if the inline storage is too small or doesn't have the sufficient alignment.

lewissbaker commented on doc/type_erasure.md:
```diff
@@ -0,0 +1,590 @@
+    // Uses std::allocator as the allocator.
+    template<typename T, typename... Args>
+      requires std::constructible_from<T, Args...>
+    explicit any_unique(std::in_place_type_t<T>, Args&&... args);
```
The use of `in_place_type_t<T>`  here is to specify that the object of type T is constructed in-place in the storage managed by the `any_unique` type, regardless of whether it's heap-allocated or not. It's not specifying that the storage is allocated "inline" inside the any_unique wrapper object.

Without this, we'd need to construct a temporary T and then move it into the storage owned by the any_unique wrapper.
This wouldn't work for objects that are not move-constructible (like an operation_state).

i.e. this is the equivalent of `emplace(args...)`  vs `push_back(T(args...))`

lewissbaker commented on doc/type_erasure.md:
```diff
@@ -0,0 +1,590 @@
+    template<typename T>
+      requires std::constructible_from<std::remove_cvref_t<T>, T> &&
+          !(std::same_as<remove_cvref_t<T>, any_unique> || instance-of-in_place_type_t<T>>)
+    /* implicit */ any_unique(T&& x);
```
I'm pretty sure the actual implementation constrains on this, I just omitted it from the documentation here because it made the requires clause overly complicated.
I can add it back here if you think it's important.

lewissbaker replied:
Actually, it seems like the current `any_unique` constructor doesn't actually constrain on the type being type-erasable...

lewissbaker commented on doc/type_erasure.md:
```diff
@@ -0,0 +1,590 @@
+    basic_any_object(basic_any_object&& other) noexcept(RequireNoexceptMove);
+
+    // Destroys the type-erased object.
+    ~basic_any_object();
```
I had left out the implementation of a move-assignment operator for now pending the discussion on potentially-throwing move-constructor and the possibility of this leaving the object in an invalid state (and thus needing an invalid vtable where every operation except destruction was `std::terminate()`)

lewissbaker commented on include/unifex/detail/any_heap_allocated_storage.hpp:
```diff
@@ -0,0 +1,164 @@
+        base(const base& other)
+#if UNIFEX_CXX_CONCEPTS
+            requires copy_constructible<T>
+#endif
```
I thought there were issues with defining the copy-constructor as a template - in that the compiler might still auto-generate a default copy-constructor for you that the compiler would choose over your template constructor?
Or is that inhibited here because we are also defining a move-constructor below?

ericniebler replied:
Right, defining a move constructor suppresses the generation of a copy constructor. Otherwise, you're correct.

lewissbaker commented on include/unifex/detail/any_heap_allocated_storage.hpp:
```diff
@@ -0,0 +1,164 @@
+            std::allocator_traits<allocator_type>::deallocate(
+                allocCopy, state_, 1);
+          }
+        }
```
This type doesn't need move assignment - if the type erasing wrapper wants to implement assignment then it will usually destroy the current value and construct a new one - it won't call through to the underlying move-assignment operator of the concrete type.

ericniebler replied:
That's fine, except that I wouldn't implement it the way you describe. That has the potential to leave the object in an invalid state, and violates invariants in the case of self-move. Instead, I would:

```c++
base& operator=(base b) {
  swap(b);
  return *this;
}
```

Obvs, you need to implement `swap` as well.

lewissbaker replied:
Self-move-assignment should be handled at the `any_object::operator=()` implementation.
It would need to handle both the inline/heap storage strategies anyway.

e.g.
```c++
any_object& operator=(any_object&& other) noexcept(RequireNothrowMove) {
  if (this != &other) {
    vtable_->template get<destroy_t>()(&storage_);
    vtable_ = empty_vtable();
    other.vtable_->template get<move_construct_t>()(&storage_, &other.storage_);
    vtable_ = other.vtable_;
  }
  return *this;
}
```

lewissbaker commented on include/unifex/any_object.hpp:
```diff
@@ -0,0 +1,240 @@
+            std::allocator_arg,
+            DefaultAllocator(),
+            std::in_place_type<T>,
+            static_cast<Args&&>(args)...) {}
```
Do you know which symbols were bloating here?
Could this be handled by a specialisation of `any_heap_allocated_storage` to simplify the case where the allocator is `std::allocator`?

lewissbaker commented on include/unifex/any_object.hpp:
```diff
@@ -0,0 +1,240 @@
+    ~type() {
+      auto* destroy = vtable_->template get<detail::_destroy_cpo>();
+      destroy(detail::_destroy_cpo{}, &storage_);
+    }
```
If we want it to be movable and not just move_constructible.

I was holding off on implementing the move-assignment until the question of how to handle invalid-states in the case that the move-constructor is potentially throwing.

lewissbaker replied:
I've now added a move-assignment operator.

lewissbaker commented on include/unifex/any_unique.hpp:
```diff
@@ -409,7 +169,10 @@ class _byval<CPOs...>::type
  private:
-  using vtable_holder_t = vtable_holder<_deallocate_cpo, CPOs...>;
+  using vtable_holder_t = std::conditional_t<
+    (sizeof...(CPOs) <= 2),
```
The idea here was to keep the `any_unique` size below 4x pointers so that it fits inside a cache-line (or half a cache-line).
We have 2 user-defined CPOs and the deallocate CPO in the vtable, plus the pointer to the object (= 4 pointers).

It is 3 for the any_ref because it doesn't need to store a deallocate entry in the vtable.

lewissbaker commented on include/unifex/swap.hpp:
```diff
@@ -30,29 +30,19 @@ template <typename T, typename U, typename = void>
-#ifdef CPP_WORKAROUND_MSVC_895622
-    nope swap();
-#endif
+    struct yes_type {};
+    struct no_type {};
```
These types don't seem to be used.

lewissbaker commented:
> > * If I pass an allocator with std::allocator_arg to the constructor but the type to be constructed fits in the small-object buffer then should it use the allocator or should it still use the small-object buffer, ignoring the allocator (currently does the latter)
> >   Should any_object allow non-movable objects that would be heap-allocated because it's a large object (move ctor is not called on concrete type in this case)?
> 
> I agree it should ignore the allocator, except to use the allocator to do the inplace construction and destruction. (I don't know why that's a thing but it is.)

Hmm, I don't think that's something I'm currently doing.
Instead I'm just ignoring the allocator and calling placement-new to construct the object inline.

I assume that people recommend doing this so that if your type is, say `std::pmr::vector<int>` and the allocator is a `std::pmr::polymorphic_allocator` that the vector will use that allocator?

Do you think this is something that can be added later?

> > * If so, should any_object also store small objects that are not movable in a heap-allocation?
> >   Or should we just always require that wrapped types are always move-constructible? (this is the current behaviour)
> 
> I agree with requiring move construction.

Ok, that just means that types like `operation_state` won't be able to be wrapped up in an `any_object`.

> > * any_object currently utilises ADL-isolation techniques but this means we can't deduce template args for this type. Is this going to break any use-cases for writing customisations of algorithms for any_object?
> 
> In my prototype for P2300, I used a different expression of the ADL isolation technique. Rather than `foo<Args...>::type`, I used `foo<id<Args>::type...>`. Seems to work well. Maybe libunifex should switch. That would be a large undertaking, though.

Hmm, interesting.
Is `id<T>::type` is some struct whose nested `::type` is `T`?

So you can write:
```c++
template<typename ArgId>
void tag_invoke(tag_t<some_cpo>, foo<ArgId>) {
  using T = typename ArgId::type;
  T::do_foo();
}
```
 
> > * Should the move constructor leave the source object in a valid but moved from state when moving a small object? Or should it destroy the source object after a successful move?
> 
> I don't think it should destroy the source object. There may be people who will expect the moved-from object to be valid.

Note that this is only for the small objects.
Large, heap-allocated objects will always leave the source object in an invalid state.
So unless you _know_ that the object is a small object and is being stored inline you can't really know whether the `any_object` is valid to use once it's in the moved-from state.

ericniebler commented:
> > > * If I pass an allocator with std::allocator_arg to the constructor but the type to be constructed fits in the small-object buffer then should it use the allocator or should it still use the small-object buffer, ignoring the allocator (currently does the latter)
> > >   Should any_object allow non-movable objects that would be heap-allocated because it's a large object (move ctor is not called on concrete type in this case)?
> > 
> > I agree it should ignore the allocator, except to use the allocator to do the inplace construction and destruction. (I don't know why that's a thing but it is.)
> 
> Hmm, I don't think that's something I'm currently doing.
> Instead I'm just ignoring the allocator and calling placement-new to construct the object inline.
> 
> I assume that people recommend doing this so that if your type is, say `std::pmr::vector<int>` and the allocator is a `std::pmr::polymorphic_allocator` that the vector will use that allocator?

The allocator's `construct` can actually add additional arguments when calling the object's constructor. So it's a correctness issue.

> Do you think this is something that can be added later?

I don't think this is something that should wait.

> > > * If so, should any_object also store small objects that are not movable in a heap-allocation?
> > >   Or should we just always require that wrapped types are always move-constructible? (this is the current behaviour)
> > 
> > I agree with requiring move construction.
> 
> Ok, that just means that types like `operation_state` won't be able to be wrapped up in an `any_object`.
> 
> > > * any_object currently utilises ADL-isolation techniques but this means we can't deduce template args for this type. Is this going to break any use-cases for writing customisations of algorithms for any_object?
> > 
> > In my prototype for P2300, I used a different expression of the ADL isolation technique. Rather than `foo<Args...>::type`, I used `foo<id<Args>::type...>`. Seems to work well. Maybe libunifex should switch. That would be a large undertaking, though.
> 
> Hmm, interesting.
> Is `id<T>::type` is some struct whose nested `::type` is `T`?
> 
> So you can write:
> ```c++
> template<typename ArgId>
> void tag_invoke(tag_t<some_cpo>, foo<ArgId>) {
>   using T = typename ArgId::type;
>   T::do_foo();
> }
> ```

Yes.

> > > * Should the move constructor leave the source object in a valid but moved from state when moving a small object? Or should it destroy the source object after a successful move?
> > 
> > I don't think it should destroy the source object. There may be people who will expect the moved-from object to be valid.
> 
> Note that this is only for the small objects.
> Large, heap-allocated objects will always leave the source object in an invalid state.

Moved-from objects shouldn't be in an invalid state, but it's ok if certain members have a precondition that the object is not moved-from.

> So unless you _know_ that the object is a small object and is being stored inline you can't really know whether the `any_object` is valid to use once it's in the moved-from state.

Hrm, good point. We better be consistent. We should destroy the source object and put the source object into some valid, null, moved-from state.

lewissbaker commented on doc/type_erasure.md:
```diff
@@ -0,0 +1,590 @@
+    shapes::height,
+    shapes::area,
+    unifex::overload<
+      any_shape(const unifex::this_, float)>(shapes::scale_by)>; // error: can't reference 'any_shape' in its own definition
```
I'll try to rework this section to introduce these APIs earlier.

Yeah, I've been wondering a bout `self` as well.
If we choose to do the rename we can do it as a separate PR, though.

lewissbaker commented:
> The allocator's `construct` can actually add additional arguments when calling the object's constructor. So it's a correctness issue.
> 
> > Do you think this is something that can be added later?
> 
> I don't think this is something that should wait.

I _think_ this needs the new C++20 `std::make_obj_using_allocator()` function to be able to do this effectively.
I'm not sure what the strategy for be for C++17, but it seems that libc++ hasn't yet implemented this function.

I'll defer working on a backport of this until a later PR (it needs all sorts of specialisations for `std::pair`, `std::tuple`, etc. to be implemented properly...).

> > > > * Should the move constructor leave the source object in a valid but moved from state when moving a small object? Or should it destroy the source object after a successful move?
> > > 
> > > 
> > > I don't think it should destroy the source object. There may be people who will expect the moved-from object to be valid.
> > 
> > 
> > Note that this is only for the small objects.
> > Large, heap-allocated objects will always leave the source object in an invalid state.
> 
> Moved-from objects shouldn't be in an invalid state, but it's ok if certain members have a precondition that the object is not moved-from.

I guess all operations other that destruction and move-assignment would fail the pre-condition that the `any_object` instance is in a moved-from state.

> > So unless you _know_ that the object is a small object and is being stored inline you can't really know whether the `any_object` is valid to use once it's in the moved-from state.
> 
> Hrm, good point. We better be consistent. We should destroy the source object and put the source object into some valid, null, moved-from state.

If we unconditionally destroy the source object then this would incur a slight runtime performance overhead as we'd need to destroy the object, replace the vtable with a dummy vtable and then the destructor of the `any_object` would still need to call the destructor vtable entry in its destructor, resulting in an extra indirect call (albeit to an empty function).

We could just say that "the source object is in a valid but moved-from state" and say that calling CPOs on the source object is invalid (regardless of whether or not it might work)?

lewissbaker commented on doc/type_erasure.md:
````diff
@@ -0,0 +1,590 @@
+```
+
+To work around this recursive definition we need to define the `any_shape` type
+as a new class that inherits from the appropriate `any_value_t` specialisation.
````
I added a definition of `any_value_t` to clarify here.

lewissbaker commented on doc/type_erasure.md:
```diff
@@ -0,0 +1,590 @@
+  unifex::overload<void(unifex::this_&&, int) noexcept>(unifex::set_value),
+
+  // Use 'sig' helper in conjunction with overload()
+  unifex::overload(unifex::set_error, unifex::sig<void(unifex::this&&, std::exception_ptr) noexcept>)
```
Removed.
